### ISSUE TITLE
Unified `AS OF` behaviors

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#2862]
We have a few issues related to AS OF in various contexts (#1210, #2803). I think they can potentially be unified up but there were a few cases that I was not entirely certain about, and that's what this musing is about.

Here are some contexts in which we might see `AS OF <time>` and I wanted to try them out. Only `TAIL` is actually implemented now.

| Context | Compaction | Query | Stream valid | Implemented |
|----------:|----:|---:|---:|--:|
| `TAIL` | forward to `time` | at `time` | from `time` | Y |
| `SELECT` | forward to `time` | at `time` |  | N |
| `CREATE SOURCE` | forward to `time` |  | from `time` | N |
| `CREATE VIEW` | forward to `time` |  | from `time` | N |
| `CREATE INDEX` | forward to `time` |  | from `time` | N |
| `CREATE SINK` | forward to `time` | at `time` | from `time` | N |

Most of these seem to make sense to me. Meaning, I think they are internally consistent and at least superficially workable. They might not be what folks expect, though.

---

One gotcha that stood out was that there is potentially a difference between
```
-- Installs <query> valid from <time>
CREATE VIEW <name> AS <query> AS OF <time>
```
and 
```
-- Should this lock <name> in at <time> for future queries?
CREATE VIEW <name> AS (SELECT <query> AS OF <time>)
```
That is, we are overloading `SELECT <query> AS OF <time>` to mean both "perform the query at this time" and "perform *and maintain* the query at this time". I could imagine folks believing that both of the two above things should have the semantics of the bottom thing.

---

The proposal has some other properties that seem interesting.

1. We get `CREATE SOURCE .. AS OF <time>` which allows people to specify when their input data should be treated as valid from. This might want to be a conversation between `materialized` and the metadata of the source, but at least we are talking about it.

2. We could imagine `ALTER VIEW <foo> WITH AS OF AS <time>` to allow users to interactively modify compaction frontiers, both advancing them and holding them back as needed.

---

I'm not 100% clear on how often these cases will get used, so it's not unreasonable to wait a bit and see what people who are more serious about temporal data processing might want. Learning a bit more, I could imagine a pile of related temporal features (e.g. `WITH FREQUENCY = 1s`, stuff like that) and it probably makes sense to look at all of these before leaping in to irrevocable action.

**Comment by JLDLaughlin** [comment]
`CREATE SINK .. AS OF` implemented with: https://github.com/MaterializeInc/materialize/pull/3467

**Comment by rjnn** [comment]
Should `DROP VIEW`, `DROP SINK`, `DROP INDEX` also support `AS OF`? Such behavior would allow one to control the termination points precisely, which would be useful to line things up when performing dropping and recreation of views to deal with schema changes.

Another feature (related, but probably separate) would be to have `DROP` commands report the times at which the `DROP` was effected, so that it could be used in the new creations.

**Comment by benesch** [comment]
There are some problems with `DROP ... AS OF` from an implementation standpoint. Dropping happens asynchronously, and may not even happen at all, I think, if there's another chunk of dataflow keeping the thing you dropped alive. Plus you'll need some synchronization, I think, in order to ensure that all workers drop at the same moment?

We definitely need a story around this for schema changes + sinks, but it's probably worth brainstorming if we can sidestep `DROP .. AS OF` somehow, because it will be hard.

**Comment by frankmcsherry** [comment]
Some thoughts here:

The main issue I can see is that the times that have been written out race with any user session in which one might type `DROP .. AS OF`. Perhaps by the time the server has gotten around to dropping things, those times have passed.

The easiest thing I think is that each sink produces correct prefixes of output changes, and so if you the user want a clean cut at `<time>` you should ensure that the frontier is complete through `<time>` and then drop the sink. You are able to filter out times greater than `<time>`, and because of races the system was unlikely to guarantee that this would happen itself.

Short version: the data probably do a better job of describing the when the computations were cleanly stopped than anything reported by the interactive session would. We should double check that sink behaves this way (that `DROP SINK` ceases producing output changes before it drops its input tokens). Ping @ruchirK for that!

**Comment by benesch** [comment]
Let's move discussion on `DROP ... AS OF` over to https://github.com/MaterializeInc/materialize/issues/3616, since I think we're jumping the gun by talking about implementation specifics before figuring out what we really need out of this feature.

**Comment by frankmcsherry** [comment]
`CREATE SINK .. AS OF` was removed in #5866. I wanted to take some notes about why so that next time around we can come at it with all the information.

Creating any object "as of" a time interacts non-trivially with catalog persistence, as when Materialize starts back up the catalog needs to understand how to re-install that object with the same "as of" property, even if the rest of the world has moved on from that time. It happens that our early "as of" methods, SELECT and TAIL don't live in the catalog, and so don't exercise this issue. CREATE SINK does, and e.g. so would CREATE SOURCE and CREATE VIEW if we supported them.

The core of the problem seems to be that we do not maintain timestamping information for the RT class of sources. Across restarts we recreate new timestamping information, and present what is essentially a new source, at least for anyone who feels strongly about timestamp information (e.g. CREATE SINK .. AS OF). Had we maintained persisted timestamp information across restarts (something I think @ruchirK is looking at) then we could just hold back compaction of that information by some function of timestamps mentioned in the catalog and frontiers we've committed as output. There is still non-trivial work to do on restart, to graft some new output onto "what we have produced already", but there are known ways to do this (start a non-snapshot sink before the last recorded time and rely on idempotence in the overlap).

Things are "easier" with CREATE SOURCE and CREATE VIEW who would just need timestamp information to remain uncompacted beyond their stated AS OF bounds.

So, these features are blocked on Materialize not being able to support them in a fault-tolerant manner. Doing so requires a bit more attention paid to recovering non-deterministic decisions Materialize has made about data and committed to publicly.
